vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc

```diff
@@ -745,25 +745,22 @@ void LedgerImpl::OnRecoverWallet(
 
 void LedgerImpl::SolveGrantCaptcha(
     const std::string& solution,
-    const std::string& promotionId) const {
-  bat_grants_->SetGrant(solution, promotionId);
+    const std::string& promotionId,
+    ledger::SolveGrantCaptchaCallback callback) {
+  auto callback_internal =
+      std::bind(&LedgerImpl::OnSolveGrantCaptcha, this, _1, _2, callback);
+  bat_grants_->SolveGrantCaptcha(solution, promotionId, callback_internal);
 }
 
-void LedgerImpl::OnGrantFinish(ledger::Result result,
-                               const braveledger_bat_helper::GRANT& grant) {
-  ledger::GrantPtr newGrant = ledger::Grant::New();
-
-  newGrant->altcurrency = grant.altcurrency;
-  newGrant->probi = grant.probi;
-  newGrant->expiry_time = grant.expiryTime;
-  newGrant->promotion_id = grant.promotionId;
-  newGrant->type = grant.type;
-
-  if (grant.type == "ads") {
+void LedgerImpl::OnSolveGrantCaptcha(
+    const ledger::Result result,
+    ledger::GrantPtr grant,
+    ledger::SolveGrantCaptchaCallback callback) {
+  if (grant->type == "ads") {
```

Comment:
check if grant is null

Reply:
and if so return and do a callback

Reply:
This one's addressed at the beginning of the function.